### ISSUE TITLE
NativeAOT publish regression in preview version (worked with alpha) hangs and uses 20GB+ memory

### ISSUE THREAD

**Issue by Beau-Gosse-dev** [dotnet/runtime#68079]
### Description

NativeAOT publish no longer works (seems to hang and leak memory) after upgrading to the preview version of the ILCompiler from the Alpha version.

### Reproduction Steps

Create a simple console app with 3 files, Program.cs, nuget.config, and ConsoleApp.csproj with contents below. Then run:
`dotnet publish -r win-x64 -c Release --self-contained`

Program.cs
```cs
using Npgsql;
var conn = new NpgsqlConnection("dummyConnectionString");
```

ConsoleApp.csproj
```xml
<Project Sdk="Microsoft.NET.Sdk">
    <PropertyGroup>
        <OutputType>Exe</OutputType>
        <TargetFramework>net7.0</TargetFramework>
        <ImplicitUsings>enable</ImplicitUsings>
    </PropertyGroup>
    <ItemGroup>
        <PackageReference Include="Npgsql" Version="5.0.7" />
        <PackageReference Include="Microsoft.DotNet.ILCompiler" Version="7.0.0-*" />
    </ItemGroup>
</Project>
```

nuget.config
```xml
<?xml version="1.0" encoding="utf-8"?>
<configuration>
 <packageSources>
    <clear />
    <add key="dotnet7" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet7/nuget/v3/index.json" />
    <add key="nuget" value="https://api.nuget.org/v3/index.json" />
 </packageSources>
</configuration>
```

### Expected behavior

Native binary is published in about 4 minutes and the ILCompiler uses about 7GB of memory (These numbers are taken from the alpha build)

### Actual behavior

It never finishes. I canceled it after 20 minutes and 20GB of memory usage. 

### Regression?

Yes it did work with the latest alpha version of the ILCompiler from the experimental nuget feed. For example, it works with ILCompiler version `7.0.0-alpha.1.22068.1`

### Known Workarounds

Go back to experimental alpha ILCompiler

### Configuration

`dotnet --version` locally = `7.0.100-preview.3.22179.4`

Target Framework is .NET 7, but seems to reproduce with Target Framework .NET 6 too. 

Tested on Windows 10, but also seems to reproduce on Mac and Linux.

x86 architecture

### Other information

With the broken version, it also spits out tons more logs like the below. This is just a small random sample of the logs:

```
ILC: Method '[Npgsql]Npgsql.TypeHandling.NpgsqlTypeHandler`1<NpgsqlTypes.NpgsqlRange`1<NpgsqlTypes.NpgsqlRange`1<NpgsqlTypes.NpgsqlRange`1<NpgsqlTypes.NpgsqlRange`1<NpgsqlTypes.NpgsqlRange`1<NpgsqlTypes.NpgsqlLogSequenceNumber>>>>>>.ValidateObjectAndGetLength(object,NpgsqlLengthCache&,NpgsqlParameter)' will always throw because: Failed to load type 'NonGenericValidateAndGetLength<NpgsqlTypes.NpgsqlRange`1<NpgsqlTypes.NpgsqlRange`1<NpgsqlTypes.NpgsqlRange`1<NpgsqlTypes.NpgsqlRange`1<NpgsqlTypes.NpgsqlRange`1<NpgsqlTypes.NpgsqlLogSequenceNumber>>>>>>' from assembly 'Npgsql, Version=5.0.7.0, Culture=neutral, PublicKeyToken=5d8b90d52f46fda7'
 
ILC: Method '[Npgsql]Npgsql.TypeHandling.NpgsqlTypeHandler`1<NpgsqlTypes.NpgsqlRange`1<NpgsqlTypes.NpgsqlRange`1<NpgsqlTypes.NpgsqlRange`1<NpgsqlTypes.NpgsqlRange`1<NpgsqlTypes.NpgsqlRange`1<NpgsqlTypes.NpgsqlPath>>>>>>.WriteWithLengthInternal<NpgsqlRange`1<NpgsqlRange`1<float64>>>(NpgsqlRange`1<NpgsqlRange`1<float64>>,NpgsqlWriteBuffer,NpgsqlLengthCache,NpgsqlParameter,bool,CancellationToken)'will always throw because: Failed to load type'<>c__DisplayClass13_0`1<NpgsqlTypes.NpgsqlRange`1<NpgsqlTypes.NpgsqlRange`1<NpgsqlTypes.NpgsqlRange`1<NpgsqlTypes.NpgsqlRange`1<NpgsqlTypes.NpgsqlRange`1<NpgsqlTypes.NpgsqlPath>>>>>, NpgsqlTypes.NpgsqlRange`1<NpgsqlTypes.NpgsqlRange`1<Double>>>' from assembly 'Npgsql, Version=5.0.7.0, Culture=neutral, PublicKeyToken=5d8b90d52f46fda7'
  
ILC: Method '[Npgsql]Npgsql.TypeHandling.NpgsqlTypeHandler`1<NpgsqlTypes.NpgsqlRange`1<NpgsqlTypes.NpgsqlRange`1<NpgsqlTypes.NpgsqlRange`1<NpgsqlTypes.NpgsqlRange`1<NpgsqlTypes.NpgsqlRange`1<NpgsqlTypes.NpgsqlLogSequenceNumber>>>>>>.CreateArrayHandler(PostgresArrayType,ArrayNullabilityMode)' will always throw because: Failed to load type Npgsql.TypeHandlers.ArrayHandler`1<NpgsqlTypes.NpgsqlRange`1<NpgsqlTypes.NpgsqlRange`1<NpgsqlTypes.NpgsqlRange`1<NpgsqlTypes.NpgsqlRange`1<NpgsqlTypes.NpgsqlRange`1<NpgsqlTypes.NpgsqlLogSequenceNumber>>>>>>' from assembly 'Npgsql, Version=5.0.7.0, Culture=neutral, PublicKeyToken=5d8b90d52f46fda7'
```

**Comment by teo-tsirpanis** [comment]
Which version of the ILCompiler package are you using? Are you sure it is the latest one? In case you don't know, [the package is now published on NuGet.org](https://www.nuget.org/packages/Microsoft.DotNet.ILCompiler). I suspect that `7.0.0-*` resolved to an older version; I remember something similar happening to myself.

**Comment by Beau-Gosse-dev** [comment]
@teo-tsirpanis Thanks for the info, I didn't realize it was on NuGet.org already. But I was able to reproduce the issue with all of these versions. It looks like the preview.3 version I tried matches what is latest on NuGet.org
```
7.0.0-preview.4.22214.3
7.0.0-preview.4.22213.8 
7.0.0-preview.3.22175.4
```

**Comment by teo-tsirpanis** [comment]
The memory blowup looks like a trimmer failure. Can you try again with the latest version of NpgSql?

**Comment by teo-tsirpanis** [comment]
[Work has been done](https://github.com/npgsql/npgsql/issues/3300) to ensure Npgsql is friendly to trimming, but its assemblies [are not marked as trimmable](https://docs.microsoft.com/en-us/dotnet/core/deploying/trimming/prepare-libraries-for-trimming#set-istrimmable).

c.c. @roji

**Comment by smhmhmd** [comment]
Same issue with  7.0.0-preview.3.22175.4

**Comment by roji** [comment]
@teo-tsirpanis yeah, there's still pending trimming work on the Npgsql side, so version 6.0 and below aren't marked as trimmable. This is ideally something I'll get done for Npgsql 7.0.

However, I notice above that Npgsql 5.0 is being used - can you please try out the latest 6.0 patch and report back?

**Comment by Beau-Gosse-dev** [comment]
@roji and @teo-tsirpanis I upgraded to the latest preview version of Npgsql `7.0.0-preview.2` and it works! It's interesting to me that the ILCompiler could handle version 5 of Npgsql but the newer ILCompiler couldn't. Is that due to more aggressive trimming or something?

**Comment by smhmhmd** [comment]
@roji 
Compile builds for me after I changed to                 
      `<PackageReference Include="Npgsql" Version="6.0.3" />`

**Comment by roji** [comment]
@Beau-Gosse-dev can you confirm that upgrading to 6.0.3 fixes the issue?

**Comment by Beau-Gosse-dev** [comment]
@roji, yes 6.0.3 also fixes the issue.

**Comment by agocke** [comment]
This sounds like it's probably https://github.com/npgsql/npgsql/issues/4057. That fix was targeted at 6.0 for npgsql, so I'll assume that the intended fix is to upgrade to that version, and that seems to work.

**Comment by smhmhmd** [comment]
@agocke @roji I hope there is a test in npgsql to build nativeaot
